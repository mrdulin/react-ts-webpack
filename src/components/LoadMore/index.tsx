import * as React from 'react';
import * as style from './style.module.scss';

interface ILoadMoreProps {
  auto?: boolean;
  onLoadMore: (cb: () => void) => void;
  hasMore: boolean;
}

interface ILoadMoreState {
  isLoading: boolean;
}

class LoadMore extends React.PureComponent<ILoadMoreProps, ILoadMoreState> {

  public static defaultProps: ILoadMoreProps = {
    auto: true,
    onLoadMore: () => {
    },
    hasMore: false
  };

  public state = {
    isLoading: false
  };

  private html: HTMLHtmlElement;
  private body: HTMLElement;

  public constructor() {
    super();
    this.onScroll = this.onScroll.bind(this);
  }

  public componentDidMount() {
    this.body = document.body;
    this.html = document.getElementsByTagName('html')[0];

    this.bindEvent(this.props);
  }

  public componentWillReceiveProps(nextProps: ILoadMoreProps) {
    this.bindEvent(nextProps);
  }

  public componentWillUnmount() {
    window.removeEventListener('scroll', this.onScroll, false);
  }

  public render(): React.ReactElement<ILoadMoreProps> {
    const {isLoading}: { isLoading: boolean } = this.state;
    const {hasMore}: { hasMore: boolean } = this.props;

    return (
      <a style={{display: hasMore && this.isBottom ? 'block' : 'none'}} 
        className={style.loadMoreRoot} onClick={this.onLoadMore}>{isLoading ? '加载中...' : '加载更多'}</a>
    );
    
  }

  private bindEvent(props: ILoadMoreProps): void {
    const {hasMore, auto} = props;
    // console.log(hasMore, auto, this.isScrollable);

    if (hasMore && auto && this.isScrollable) {
      window.addEventListener('scroll', this.onScroll, false);
    } else {
      window.removeEventListener('scroll', this.onScroll, false);
    }
  }

  private toggleLoading(isLoading: boolean) {
    this.setState({isLoading});
  }

  private onLoadMore(): void {
    const auto: boolean | undefined = this.props.auto;

    if (!auto || this.state.isLoading) {
      return;
    }

    this.toggleLoading(true);

    this.props.onLoadMore(this.onAfterLoadMore);
  }

  private onAfterLoadMore = (): void => {
    console.log('onAfterLoadMore');
    this.toggleLoading(false);
  }

  get isScrollable(): boolean {
    return this.body.scrollHeight > this.html.clientHeight;
  }

  get isBottom(): boolean {
    if (!this.html) {
      return false;
    }
    return this.html.clientHeight + this.body.scrollTop >= this.body.scrollHeight;
  }

  private onScroll(e: Event): void {
    console.log('isBottom', this.isBottom);
    if (this.isBottom) {
      this.onLoadMore();
    }
  }

}

export default LoadMore;
