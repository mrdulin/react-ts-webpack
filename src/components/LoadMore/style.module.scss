.load-more-root{
  display: block;
  font-size: 14px;
  color: #666;
  text-align: center;
  padding: 10px;
}
