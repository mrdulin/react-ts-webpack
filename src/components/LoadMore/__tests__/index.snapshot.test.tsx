/**
 * Created by dulin on 2017/7/6.
 */
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import LoadMore from '../';

interface IRenderer extends renderer.Renderer {
  getInstance(): any;
}

const create = (nextElement: React.ReactElement<any>, options?: renderer.TestRendererOptions): IRenderer => {
  return renderer.create(nextElement, options) as IRenderer;
};

describe('component - LoadMore test suites', () => {

  const onLoadMore = (cb: () => void): void => {
    console.log('onLoadMore');
  };

  it('renders correctly when isBottom=true, hasMore=true, isLoading=true', () => {

    const component: IRenderer = create(
      <LoadMore hasMore={true} onLoadMore={onLoadMore}/>
    );

    // const componentInstance = component.getInstance();
    // Object.defineProperty(componentInstance, 'isBottom', {
    //   get: jest.fn(() => true)
    // });
    // componentInstance.setState({isLoading: true});
    //
    // console.log(componentInstance.state);
    // expect(componentInstance.props.hasMore).toBeTruthy();
    // expect(componentInstance.isBottom).toBeTruthy();
    const tree = component.toJSON();
    
    console.log(tree);
    // expect(tree.props.hasMore).toBeTruthy();
    
    expect(tree).toMatchSnapshot();

  });
  
  // it('does not render when isBottom=false, hasMore=false, isLoading=false', () => {
  //
  //   const component: IRenderer = create(
  //     <LoadMore hasMore={false} onLoadMore={onLoadMore}/>
  //   );
  //   expect(component.toJSON()).toMatchSnapshot();
  //
  // });
  //
  // it('does not render when isBottom=false, hasMore=true, isLoading=false', () => {
  //  
  //   const component: IRenderer = create(
  //     <LoadMore hasMore={false} onLoadMore={onLoadMore}/>
  //   );
  //   expect(component.toJSON()).toMatchSnapshot();
  // });
  //
  // it('renders correctly when isBottom=true, hasMore=true, isLoading=false', () => {
  //  
  //   const component: IRenderer = create(
  //     <LoadMore hasMore={true} onLoadMore={onLoadMore}/>
  //   );
  //  
  //   const componentInstance = component.getInstance();
  //   Object.defineProperty(componentInstance, 'isBottom', {
  //     get: jest.fn(() => true)
  //   });
  //
  //   expect(componentInstance.isBottom).toBeTruthy();
  //   expect(componentInstance.props.hasMore).toBeTruthy();
  //   expect(component.toJSON()).toMatchSnapshot();
  //  
  //  
  // });
});
