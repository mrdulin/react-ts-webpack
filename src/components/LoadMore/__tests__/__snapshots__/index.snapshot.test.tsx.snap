// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component - LoadMore test suites renders correctly when isBottom=true, hasMore=true, isLoading=true 1`] = `
<a
  className="loadMoreRoot"
  onClick={[Function]}
  style={
    Object {
      "display": "none",
    }
  }
>
  加载更多
</a>
`;
